src/main/webapp/js/script.js

```diff
@@ -0,0 +1,226 @@
+ * @param url - the url of the YouTube video
+ * @returns a promise which upon success returns a JSON 
+ *         string encoding the captions and timestamps 
+ */
+function getCaptions(trackId, url){
+    return new Promise((success, failure) => {
+        gapi.client.youtube.captions.download({
+            "id": trackId,
+            "tlang": "en",
+            "tfmt": "sbv"
+        }).then(function(response){
+            parseCaptionsIntoJson(response, url).then(json => {
+                success(json);  
+            });
+        }, function(err) { 
+            console.error("errors getting captions", err); 
+            if (err.status == 403) {
+                console.error("Video has private captions!");
+            }
+            failure(err);
+        });
+    }).catch(function(error) {
+        // throw for top level handler to handle
+        throw error;   
+    });
+}
+
+/**
+ * Gets the video id from a YouTube url
+ * @param url - a YouTube video url
+ * @returns - the `id` for a YouTube `url`
+ */
+function getIdFromUrl(url) {
+    var video_id = url.split('v=')[1];
+    var ampersandPosition = video_id.indexOf('&');
+    if(ampersandPosition != -1) {
+        video_id = video_id.substring(0, ampersandPosition);
+    }
+
+    return video_id;
+}
+
+/**
+ * Excecutes a request to download captions for a YouTube 
+ * video with a given `url`, and send data to the backend servlet
+ * @param url - a url for a YouTube video
+ * @requires - an authenticated user
+ */
+// Make sure the client is loaded and sign-in is complete before calling this method.
+function execute(url) {
+    // user inputs YouTube video URL
+    const videoId = getIdFromUrl(url);
+
+    return gapi.client.youtube.captions.list({
+      "videoId": videoId,
+      "part": [
+        "id"
+      ]
+    }).then(function(response) {
+        if (response.result != null & response.result.items != null &
+            response.result.items.length > 0) {
+            const trackId = response.result.items[0].id;
+
+            getCaptions(trackId, url).then(json => {
+                // send to backend
+                console.log("final destination", json);
+                sendJsonForm(json);
+            });
+        }
+    }, function(err) {
+        // top level error handler
+        console.error("Execute error", err); 
+    });
+}
+
+/**
+ * Create and send a form with the captions JSON. May need 
+ * to authenticate users here too to prevent malicious requests.
+ * @param json - the JSON string representing the 
+ *               parsed captions response
+ */
+function sendJsonForm(json) {
+    /** 
+    var form = document.createElement('form');
+    form.method = 'POST';
+    form.action = '/caption';
+
+    var jsonInput = document.createElement('input');
+    jsonInput.type = 'hidden';
+    jsonInput.name = 'json';
+    jsonInput.value = json;
+    form.appendChild(jsonInput);
+
+    document.body.appendChild(form);
+    form.submit();
+    */
```

Comment:
I think I commented this, but if it is no longer needed for reference, it can be deleted.